US-8.1: Add Friends

cristian-aldea opened the pull request:
closes #273

Note:
- Some e2e tests were failing. I fixed them all (aside from US-2.8 which is sometimes flaky for me). they should all pass if you run them from a freshly seeded DB BEFORE 11:45pm

How to test
- Go to your profile page
- Go to "Add Friends"
- Make sure searching for strangers, accepting friend requests, and canceling pending requests works.

cristian-aldea committed:
fix: Set standard styling for h1

- Set h1 to be more visually distinguishable
- Adjusting some styling on other pages

ref #273

cristian-aldea committed:
feat(relations): Add required backend queries

ref #273

cristian-aldea committed:
feat(add friends): Implement add friends page

ref #273

cristian-aldea committed:
fix: Address linting issues

ref #273

cristian-aldea committed:
fix: update snapshots

ref #273

cristian-aldea committed:
feat: Add tests for backend

ref #273

cristian-aldea committed:
fix(e2e): Implement e2e and fix existing

ref #273

cindyslittleplanet reviewed (commented):
Everything works well in terms of functionality for me! 
I found a few linting issues and 1 failing snapshot, so make sure to correct those. 

Also, just a few suggestions:
1) If there are no requests, we should either see a message like "No added friends/No requests" or just make the header disappear since it looks a bit awkward right now. 
![image](https://user-images.githubusercontent.com/38992971/113587087-50832000-95fc-11eb-9410-beea90c9f9b6.png)

2) Another small thing is in the `user-profile`, the number next to Add Friends should match up with the number of friend requests received (Added Me), and if there isn't any friend requests, it should be hidden. Right now, it's a hardcoded 4.
![image](https://user-images.githubusercontent.com/38992971/113587278-8d4f1700-95fc-11eb-9254-34493bfd5b88.png)

osahtout commented on apps/mull-api/src/database/seeders/database.seeder.ts:
```diff
@@ -47,15 +47,29 @@ const US3_1Seeder = async (connection: Connection) => {
 export default class DatabaseSeeder implements Seeder {
   public async run(factory: Factory, connection: Connection): Promise<any> {
-    await factory(Event)().createMany(10);
+    await factory(Event)().createMany(20);
```
Why add more events? this will create more than 500 users

cristian-aldea replied:
My bad, I was testing something and forgot to revert

osahtout commented on apps/mull-api/src/database/seeders/database.seeder.ts:
```diff
@@ -47,15 +47,29 @@ const US3_1Seeder = async (connection: Connection) => {
+    await factory(Event)().createMany(20);
     await createPosts(connection, factory);
     await US3_1Seeder(connection);
+    await US8_1Seeder(connection);
```
I'm not sure who started this, but I don't like these names. At least change the name to the one you made to what it actually does.

cristian-aldea replied:
Well what it does is setup some users to be used for e2e tests. I think the name is fine. Do you have another name you can suggest?

osahtout commented on apps/mull-api/src/app/user/user.service.ts:
```diff
@@ -136,6 +136,24 @@ export class UserService {
+         AND INSTR(user.name, :searchInput) > 0`,
+        { id, searchInput }
+      )
+      .getMany();
```
is it possible that the id is not in the table? will it still return all the users?

cristian-aldea replied:
I'm not sure what's your question, lets talk about it today when you have the time

osahtout commented on apps/mull-ui/src/app/components/contact-row/contact-row.tsx:
```diff
@@ -27,14 +29,15 @@ export const ContactRow = ({
-    <div className="contact-container">
-      <Link to={`/user/${user.id}`}>
+    <div className="contact-container" id={id}>
+      <Link to={ROUTES.USER_BY_ID.replace(':id', user.id + '')}>
```
Why not have a constants USER = "/user/" and concatenate the id or have a method that does it instead replacing the string everywhere.

cristian-aldea replied:
In this specific case the :id is at the end. This might not always be the case. Why have a separate `"/user/"` variable when I can express where I want the user to go with the routes we already have defined?

osahtout commented on apps/mull-ui/src/app/pages/profile/add-friends/add-friends.tsx:
```diff
@@ -1,44 +1,152 @@
+  };
+
+  if (relData.loading || relData.error) {
+    return <div>something happened: {relData.error}</div>;
```
so if it's loading it will display and error?

cristian-aldea replied:
Sorry about that, I kinda left this part intact after quickly writing something so I could continue working without errors. Will be changed to be a standard loading div.

cristian-aldea committed:
fix: Address pr comments

- Add message when there are no pending requests to and from other users
- Update old placeholder code for checking for query loading and error
- Implemented proper count for pending requests to current user on user profile. Also reworked styling for count

ref #273

cristian-aldea committed:
fix: Address all test, lint and e2e issues

ref #273

cristian-aldea committed:
fix: cleanup

I'm a bit tired, forgot to remove a log and saw some typo with the casing for the text of a button

ref #273

cristian-aldea committed:
fix: rename seeder methods

ref #273

cindyslittleplanet reviewed (approved):
LGTM! 🚀 Everything passes! 

Small point, can you add back the spacing between the button `My Friends` and the sub-title `Misc.`
![image](https://user-images.githubusercontent.com/38992971/113649208-0598f500-965c-11eb-9b1f-25c6ec560690.png)

**Acceptance Criteria**

- [x] User Acceptance Flow is respected
- [x] There must exist an automated Cypress acceptance test which tests the user acceptance flow
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every new component and function
- [x] The application should still work with seeded data
- [x] Must match mockups:
   - "[Added Me + Pending Requests](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=1596%3A5062)" view
   - "[Pending Request modal](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=1695%3A83663)" view
   - "[Search Users](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=1695%3A86144)" view

cristian-aldea committed:
fix: Remove leftover styling

ref #273

ys-lin reviewed (approved):
**Acceptance Criteria**

- [x] User Acceptance Flow is respected
- [x] There must exist an automated Cypress acceptance test which tests the user acceptance flow
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every new component and function
- [x] The application should still work with seeded data
- [x] Must match mockups:
   - "[Added Me + Pending Requests](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=1596%3A5062)" view
   - "[Pending Request modal](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=1695%3A83663)" view
   - "[Search Users](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=1695%3A86144)" view
